File: PlantCompanion/Core/Domain/WateringPlanner.swift
import Foundation

/// Turns a `CareProfile` into a dated, explainable soil check.
///
/// A pure value type over a fixed calendar, so it can be exercised directly by the shared
/// contract vectors and used from any concurrency domain.
struct WateringPlanner: Sendable {
    var calendar: Calendar

    init(calendar: Calendar = .autoupdatingCurrent) {
        self.calendar = calendar
    }

    /// The interval and the phrases explaining every modifier that moved it.
    struct AppliedInterval: Hashable, Sendable {
        let days: Int
        let factors: [String]
    }

    func interval(for profile: CareProfile, asOf date: Date) -> AppliedInterval {
        let month = calendar.component(.month, from: date)
        let modifiers: [(days: Int, factor: String?)] = [
            CareRules.environmentModifier(for: profile.environment).map { ($0.days, $0.factor) },
            CareRules.lightModifier(for: profile.light).map { ($0.days, $0.factor) },
            CareRules.seasonModifier(month: month, environment: profile.environment).map { ($0.days, $0.factor) }
        ].compactMap { $0 }

        var days = max(1, profile.baselineWateringDays)
        var factors: [String] = []

        for modifier in modifiers {
            days += modifier.days
            if let factor = modifier.factor { factors.append(factor) }
        }

        days = min(max(days, CareRules.minimumIntervalDays), CareRules.maximumIntervalDays)
        return AppliedInterval(days: days, factors: factors)
    }

    func plan(for profile: CareProfile, plantID: UUID, asOf date: Date = .now) -> CareRecommendation {
        let applied = interval(for: profile, asOf: date)
        let dueDate = calendar.date(byAdding: .day, value: applied.days, to: profile.anchor) ?? date
        let dayDifference = calendar.dateComponents(
            [.day],
            from: calendar.startOfDay(for: date),
            to: calendar.startOfDay(for: dueDate)
        ).day ?? 0

        return CareRecommendation(
            plantID: plantID,
            dueDate: dueDate,
            intervalDays: applied.days,
            status: status(forDayDifference: dayDifference),
            title: title(forDayDifference: dayDifference),
            reason: "\(CareRules.Phrasing.explanation(factors: applied.factors)) \(CareRules.Phrasing.reasonSuffix)"
        )
    }

    private func status(forDayDifference difference: Int) -> CareRecommendation.Status {
        if difference < 0 { return .overdue }
        if difference == 0 { return .dueToday }
        return .upcoming
    }

    private func title(forDayDifference difference: Int) -> String {
        if difference < 0 { return CareRules.Phrasing.overdueTitle }
        if difference == 0 { return CareRules.Phrasing.dueTodayTitle }
        return CareRules.Phrasing.upcomingTitle(days: difference)
    }
}

File: ReactMockup/src/domain/wateringPlanner.ts
import { bounds, environmentModifiers, lightModifiers, phrasing, seasonModifiers } from './careRules';
import { addDays, dayDifference } from './dates';
import type {
  CareRecommendation,
  LightLevel,
  PlantEnvironment,
  RecommendationStatus,
  UserPlant,
} from './types';

/**
 * Everything the planner needs, and nothing about how a plant is stored. Keeping this
 * separate from UserPlant is what lets the golden contract vectors exercise the planner
 * without inventing a whole persisted plant for each case.
 */
export interface CareProfile {
  baselineWateringDays: number;
  environment: PlantEnvironment;
  light: LightLevel;
  anchor: Date;
}

interface AppliedInterval {
  days: number;
  factors: string[];
}

export const latestWateringDate = (plant: UserPlant): Date | undefined => {
  const timestamps = plant.careEvents
    .filter((event) => event.kind === 'watered')
    .map((event) => new Date(event.timestamp).getTime());
  return timestamps.length > 0 ? new Date(Math.max(...timestamps)) : undefined;
};

/** A plant's schedule is anchored to its last watering, or to the day it was added. */
export const careProfileFor = (plant: UserPlant): CareProfile => ({
  baselineWateringDays: plant.baselineWateringDays,
  environment: plant.environment,
  light: plant.light,
  anchor: latestWateringDate(plant) ?? new Date(plant.dateAdded),
});

export function resolveInterval(profile: CareProfile, now: Date): AppliedInterval {
  const month = now.getMonth() + 1;
  const applied = [
    environmentModifiers.find((entry) => entry.key === profile.environment),
    lightModifiers.find((entry) => entry.key === profile.light),
    // seasonModifiers is ordered by precedence, so the first match is the only one that applies.
    seasonModifiers.find((entry) =>
      entry.months.includes(month) && !entry.excludesEnvironments.includes(profile.environment)),
  ];

  let days = Math.max(1, profile.baselineWateringDays);
  const factors: string[] = [];

  for (const modifier of applied) {
    if (!modifier) continue;
    days += modifier.days;
    if (modifier.factor) factors.push(modifier.factor);
  }

  return {
    days: Math.min(Math.max(days, bounds.minimumIntervalDays), bounds.maximumIntervalDays),
    factors,
  };
}

export function wateringRecommendation(plant: UserPlant, now = new Date()): CareRecommendation {
  return { plantId: plant.id, ...planCare(careProfileFor(plant), now) };
}

export function planCare(profile: CareProfile, now = new Date()): Omit<CareRecommendation, 'plantId'> {
  const { days, factors } = resolveInterval(profile, now);
  const dueDate = addDays(profile.anchor, days);
  const difference = dayDifference(now, dueDate);

  return {
    dueDate: dueDate.toISOString(),
    intervalDays: days,
    status: statusFor(difference),
    title: titleFor(difference),
    reason: `${explanationFor(factors)} ${phrasing.reasonSuffix}`,
  };
}

const statusFor = (difference: number): RecommendationStatus =>
  difference < 0 ? 'overdue' : difference === 0 ? 'dueToday' : 'upcoming';

const titleFor = (difference: number): string => {
  if (difference < 0) return phrasing.overdueTitle;
  if (difference === 0) return phrasing.dueTodayTitle;
  return phrasing.upcomingTitleTemplate
    .replace('{days}', String(difference))
    .replace('{plural}', difference === 1 ? '' : 's');
};

const explanationFor = (factors: string[]): string =>
  factors.length === 0
    ? phrasing.explanationWithoutFactors
    : phrasing.explanationTemplate.replace('{factors}', factors.join(', '));
